doc/http-server.1
.TH http-server 1 "April 2020" GNU "http-server man page"

.SH NAME
http-server \- a simple zero-configuration command-line http server

.SH SYNOPSIS
.B http-server
[\fIPATH\fR]
[\fIOPTIONS\fR]

.SH DESCRIPTION
\fBhttp-server\fR is a simple, zero-configuration command-line http server. It is powerful enough for production usage, but it's simple and hackable enough to be used for testing, local development, and learning.

.SH OPTIONS

.TP
.BI [\fIPATH\fR]
The directory to serve.
Defaults to ./public if it exists, and ./ otherwise.

.TP
.BI \-p ", " \-\-port " " \fIPORT\fR
Port to use. If 0, look for the first available port, starting at 8080.
Default is 8080.

.TP
.BI \-a " " \fIADDRESS\fR
Address to use.
Default is 0.0.0.0.

.TP
.BI \-d
Show directory listings.
Default is true.

.TP
.BI \-d
Whether -d should override magic 404.html
Default is false.

.TP
.BI \-i
Display autoIndex.
Default is true.

.TP
.BI \-\-sort " " \fISORT\fR
Sort directory listings by name, modified time, or creation time.
Accepted values are name, modified, mtime, created, and birthtime.
Date sorting shows newest entries first.
Default is name.

.TP
.BI \-g ", " \-\-gzip
Serve gzip files when possible.
Default is false.

.TP
.BI \-b ", " \-\-brotli
Serve brotli files when possible.
If both brotli and gzip are enabled, brotli takes precedence.
Default is false.

.TP
.BI \-\-force\-content\-encoding
When using --gzip or --brotli, includes the content encoding
header even when the extension for the compressed file is
specified in the URL. "test.png.br" will be served the same
way as "test.png".

.TP
.BI \-e ", " \-\-ext " " \fIEXTENSION\fR
Default file extension is none is provided.

.TP
.BI \-s ", " \-\-silent
Suppress log messages from output.

.TP
.BI \-n ", " \-\-no-panic
Gracefully shut down whenever a fatal error occurs, sending stack to log file, not console.

.TP
.BI \-\-coop " " [\fIMODE\fR]
Enable COOP via the "Cross-Origin-Opener-Policy" header and sets
the "Cross-Origin-Embedder-Policy" header to "require-corp".
Optionally provide COOP mode which defaults to "same-origin".

.TP
.BI \-\-cors " " [\fIHEADERS\fR]
Enable CORS by setting "Access-Control-Allow-Origin" to "*".
Optional comma-separated headers list adds to "Access-Control-Allow-Headers".
Default Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, Range.

.TP
.BI \-H ", " \-\-header " " \fIHEADER\fR
Add an extra response header (can be used several times)

.TP
.BI \-\-private-network-access
Enable Private Network Access via the "Access-Control-Allow-Private-Network" header.

.TP
.BI \-o " " [\fIPATH\fR]
Open default browser window after starting the server.
Optionally provide a URL path to open the browser window to.

.TP
.BI \-c " " \fITIME\fR
Cache time (max-age) in seconds.
To disable caching, use \-c \-1.
Default is 3600.

.TP
.BI \-t " " \fITIMEOUT\fR
Connection timeout in seconds, e.g. -t60 for 1 minute.
To disable timeout, use \-t0.
Default is 120.

.TP
.BI \-T ", " \-\-title " " \fITITLE\fR
Custom title suffix for the terminal window.
The terminal title will be "http-server PORT [TITLE]".

.TP
.BI \-U ", " \-\-utc
Use UTC time format in log messages.

.TP
.BI \-\-log\-ip
Enable logging of the client IP address.

.TP
.BI \-P ", " \-\-proxy
Fallback proxy if the request cannot be resolved.

.TP
.BI \-\-proxy\-all
Forward every request to the proxy target and disable local file serving.
Requires \-\-proxy.

.TP
.BI \-\-proxy\-options
Pass proxy options using nested dotted objects.

.TP
.BI \-\-proxy\-config
Pass in .json configuration file.

.TP
.BI \-\-user ", " \-\-username " " \fIUSERNAME\fR
Username for basic authentication.
Can also be specified with the environment variable NODE_HTTP_SERVER_USERNAME.
Defaults to none.

.TP
.BI \-\-password " " \fIPASSWORD\fR
Password for basic authentication.
Can also be specified with the environment variable NODE_HTTP_SERVER_PASSWORD.
Defaults to none.

.TP
.BI \-S ", " \-\-tls ", " \-\-ssl
Enable https.

.TP
.BI \-C ", " \-\-cert " " [\fIFILE\fR]
Path to SSL certificate file.
If not specified, uses cert.pem.

.TP
.BI \-K ", " \-\-key " " [\fIFILE\fR]
Path to SSL key file.
If not specified, uses key.pem.
Passphrase will be read from NODE_HTTP_SERVER_SSL_PASSPHRASE (if set)

.TP
.BI \-r ", " \-\-robots " " [\fIUSER\-AGENT\fR]
Respond to /robots.txt request.
If not specified, uses "User-agent: *\\nDisallow: /]"

.TP
.BI \-\-no\-dotfiles
Do not show dotfiles.

.TP
.BI \-\-hide\-permissions
Do not show file permissions.

.TP
.BI \-\-allowed\-hosts
Comma-separated list of hosts allowed to access the server. e.g.: \-\-allowed\-hosts localhost,example.com

.TP
.BI \-h ", " \-\-help
Print usage and exit.

.TP
.BI \-v ", " \-\-version
Print version and exit.

.SH FILES
.B index.html
will be served as the default file to any directory requests.

.B 404.html
will be served if a file is not found. This can be used for SPA hosting to serve the entry page.

.SH COPYING
Copyright (c) 2011-2022 Charlie Robbins, Marak Squires, and the Contributors.
Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

.SH VERSION
Version 0.12.2


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

## [Unreleased](https://github.com/http-party/http-server/compare/v13.0.2...HEAD)

### Merged Pull Requests
#### Adds

- Add Contributing guide [`#752`](https://github.com/http-party/http-server/pull/752)
- add passphrase option [`#746`](https://github.com/http-party/http-server/pull/746)
- Add and configure release-drafter [`#750`](https://github.com/http-party/http-server/pull/750)
- add ability to pass proxyOptions [`#688`](https://github.com/http-party/http-server/pull/688)
- Adds version number to server startup output [`#734`](https://github.com/http-party/http-server/pull/734)

#### Updates

- Update some dependencies [`#740`](https://github.com/http-party/http-server/pull/740)

#### Misc

- Eslint config - replace common-style with eslint-config-populist [`#744`](https://github.com/http-party/http-server/pull/744)
- Fix CORS option detection [`#748`](https://github.com/http-party/http-server/pull/748)
- fix crash on redirect with formfeed in URL [`#749`](https://github.com/http-party/http-server/pull/749)
- Fixes --proxy without a protocol throwing an uncaught error [`#742`](https://github.com/http-party/http-server/pull/742)
- Make --ssl an alias for --tls [`#747`](https://github.com/http-party/http-server/pull/747)
- Drop Node.js 10 support [`#739`](https://github.com/http-party/http-server/pull/739)
- Replace mkdirp in tests with native JS [`#743`](https://github.com/http-party/http-server/pull/743)
- Fix tests EACCESS by finding an open port every time [`#741`](https://github.com/http-party/http-server/pull/741)
- Implement displaying last modified date in index [`#737`](https://github.com/http-party/http-server/pull/737)
- Use relative paths in directory listing #661 [`#732`](https://github.com/http-party/http-server/pull/732)

### Closed Issues
- add contributing [`#417`](https://github.com/http-party/http-server/issues/417)
- get a good port every time [`#723`](https://github.com/http-party/http-server/issues/723)
- Fixes #451 [`#451`](https://github.com/http-party/http-server/issues/451)

### Commits

- replace common-style with eslint-config-populist [`e692640`](https://github.com/http-party/http-server/commit/e692640a6eeab9b17f44f7949f4fad050cf9e804)
- update lock [`b927279`](https://github.com/http-party/http-server/commit/b927279d49c41379fcb8a97e644564a7253fda4e)
- Added auto-changelog library for CHANGELOG generation [`78f71f0`](https://github.com/http-party/http-server/commit/78f71f09033f53928db49fd50c2d0239c8dd8184)
- chore: update auto-changelog version to latest [`992cc53`](https://github.com/http-party/http-server/commit/992cc532660fc57571b6ed2a2985db4555b3cbc7)
- feat: Add changelog generation to version script [`fd86c2d`](https://github.com/http-party/http-server/commit/fd86c2ddb6ab55f0c96a92e52128864faf7f3089)
- Generated a CHANGELOG.md [`ed1fd0e`](https://github.com/http-party/http-server/commit/ed1fd0e3633527d00ecc29bfcf70642662575b5c)
- Generate updated changelog [`0689f68`](https://github.com/http-party/http-server/commit/0689f68472ee909004d001f6bc8ab3af7edf8f9b)
- Create CODE_OF_CONDUCT.md [`7c7c598`](https://github.com/http-party/http-server/commit/7c7c5987243d548be1ba26c64f4df047ff7c7c6c)
- Cors fixes [`964725e`](https://github.com/http-party/http-server/commit/964725e31c92cdf7f754ce0f784ef054d08120e3)
- update issue templates [`622302e`](https://github.com/http-party/http-server/commit/622302e5207f0179306248f5964f211d90c12808)
- use portfinder in cli test [`6f49089`](https://github.com/http-party/http-server/commit/6f4908986249818ad66a80e5721bd8538947247b)
- add release-drafter [`0d3783e`](https://github.com/http-party/http-server/commit/0d3783e0222871d105c9797ff1cd81b044261c32)
- apply minimal eslint rule overrides [`46af9d0`](https://github.com/http-party/http-server/commit/46af9d01b9662c4d65d855ff6392f340e820f504)
- upgrade basic-auth [`acf6095`](https://github.com/http-party/http-server/commit/acf60950252b49a2f2d92e48d09de73fee464303)
- make ssl an alias of tls [`bdf879a`](https://github.com/http-party/http-server/commit/bdf879aafa1389d2ae67fde1e2503161e8853a00)
- minor upgrades [`74ddfda`](https://github.com/http-party/http-server/commit/74ddfdab072051d054f155e5f66aceaba2b9b7c8)
- fix errors and formatting in shim [`7c510cf`](https://github.com/http-party/http-server/commit/7c510cf95cf1eb10bbd210664d32350df250f61e)
- Update PULL_REQUEST_TEMPLATE [`0b754bf`](https://github.com/http-party/http-server/commit/0b754bfc87e67cf7b6a3a3505b9e2e1537bc3a66)
- replace mkdirp with native js [`7025474`](https://github.com/http-party/http-server/commit/70254742a68e889594a00bc27a7793ead3b6154d)
- update url-join [`910a469`](https://github.com/http-party/http-server/commit/910a46908dbd4a0d1cc2692112c185e85d4ae7c7)
- doc: Adds documentation on auto-changelog use [`a2cc4dc`](https://github.com/http-party/http-server/commit/a2cc4dc35d7d854ea98763195a27eb45eaf7dfc3)
- update contribution guide [`49e6608`](https://github.com/http-party/http-server/commit/49e660876339a26c1ec0e5e1b6366294067f7dab)
- Added test for issue #451 [`e8db8dc`](https://github.com/http-party/http-server/commit/e8db8dc7776abb219579f3632c3bc6ae49b73bd5)
- Made proxy url checker more general [`0ba1cb3`](https://github.com/http-party/http-server/commit/0ba1cb3df185aac135a1b1baaccd7fbf9f52b102)
- Update test/cli.test.js [`046fefa`](https://github.com/http-party/http-server/commit/046fefa1402ba84aff75997de66aedd99f743372)
- Update release-drafter.yml [`a7d08e1`](https://github.com/http-party/http-server/commit/a7d08e1f9637c7cc8aae0ea5b4e6b36e51f42f07)
- update contribution guide [`df5f0aa`](https://github.com/http-party/http-server/commit/df5f0aa3d3658a5869aa684e37b000241f37d34c)

lib/http-server.js
'use strict';

var fs = require('fs'),
  union = require('union'),
  httpServerCore = require('./core'),
  auth = require('basic-auth'),
  httpProxy = require('http-proxy'),
  corser = require('corser'),
  secureCompare = require('secure-compare');
var { minimatch } = require('minimatch');

//
// Remark: backwards compatibility for previous
// case convention of HTTP
//
exports.HttpServer = exports.HTTPServer = HttpServer;

/**
 * Returns a new instance of HttpServer with the
 * specified `options`.
 */
exports.createServer = function (options) {
  return new HttpServer(options);
};

/**
 * Constructor function for the HttpServer object
 * which is responsible for serving static files along
 * with other HTTP-related features.
 */
function HttpServer(options) {
  options = options || {};
  var proxyAll = options.proxyAll === true || options.proxyAll === 'true';

  if (proxyAll && typeof options.proxy !== 'string') {
    throw new Error('proxyAll option requires "proxy" to be configured');
  }

  if (options.root) {
    this.root = options.root;
  } else {
    try {
      // eslint-disable-next-line no-sync
      fs.lstatSync('./public');
      this.root = './public';
    } catch (err) {
      this.root = './';
    }
  }

  // CRLF injection prevention
  for ( const [key, value] of Object.entries(options.headers || {}) ) {
    if (typeof key !== 'string' || typeof value !== 'string') {
      throw new Error('Header is not a string or contains CRLF');
    }
    if (key.includes('\r') || key.includes('\n') || value.includes('\r') || value.includes('\n')) {
      throw new Error('Header is not a string or contains CRLF');
    }
  }

  this.headers = options.headers || {};
  this.headers['Accept-Ranges'] = 'bytes';

  this.cache = (
    // eslint-disable-next-line no-nested-ternary
    options.cache === undefined ? 3600 :
    // -1 is a special case to turn off caching.
    // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#Preventing_caching
      options.cache === -1 ? 'no-cache, no-store, must-revalidate' :
        options.cache // in seconds.
  );
  this.showDir = options.showDir !== 'false';
  this.dirOverrides404 = options.dirOverrides404;
  this.autoIndex = options.autoIndex !== 'false';
  this.showDotfiles = options.showDotfiles;
  this.hidePermissions = options.hidePermissions;
  this.gzip = options.gzip === true;
  this.brotli = options.brotli === true;
  this.forceContentEncoding = options.forceContentEncoding === true;
  if (options.ext) {
    this.ext = options.ext === true
      ? 'html'
      : options.ext;
  }
  this.contentType = options.contentType ||
    (this.ext === 'html' ? 'text/html' : 'application/octet-stream');

  var before = options.before ? options.before.slice() : [];

  if (options.logFn) {
    before.push(function (req, res) {
      options.logFn(req, res);
      res.emit('next');
    });
  }

  if (options.username || options.password) {
    if (!options.username || !options.password) {
      throw new Error('Basic authentication requires both username and password to be specified');
    }

    before.push(function (req, res) {
      var credentials = auth(req);

      // We perform these outside the if to avoid short-circuiting and giving
      // an attacker knowledge of whether the username is correct via a timing
      // attack.
      if (credentials) {
        // if credentials is defined, name and pass are guaranteed to be string
        // type
        var usernameEqual = secureCompare(options.username.toString(), credentials.name);
        var passwordEqual = secureCompare(options.password.toString(), credentials.pass);
        if (usernameEqual && passwordEqual) {
          return res.emit('next');
        }
      }

      res.statusCode = 401;
      res.setHeader('WWW-Authenticate', 'Basic realm=""');
      res.end('Access denied');
    });
  }

  if (options.allowedHosts) {
    before.push(function (req, res) {
      let host = req.headers && req.headers.host;
      if (host) {
        // don't include port number in host check
        host = host.split(':')[0];
      }

      if (!host || !options.allowedHosts.includes(host)) {
        res.statusCode = 403;
        res.end('Access denied');
        return;
      }

      return res.emit('next');
    });
  }

  if (options.coop) {
    this.headers['Cross-Origin-Opener-Policy'] = options.coopHeader || 'same-origin';
    this.headers['Cross-Origin-Embedder-Policy'] = 'require-corp';
  }

  // CORS configuration:
  // --cors enables CORS by setting Access-Control-Allow-Origin to '*'
  // --cors=header1,header2 also adds custom headers to Access-Control-Allow-Headers
  if (options.cors) {
    this.headers['Access-Control-Allow-Origin'] = '*';
    this.headers['Access-Control-Allow-Headers'] = 'Origin, X-Requested-With, Content-Type, Accept, Range';
    if (options.corsHeaders) {
      options.corsHeaders.split(/\s*,\s*/)
        .forEach(function (h) { this.headers['Access-Control-Allow-Headers'] += ', ' + h; }, this);
    }
    before.push(corser.create(options.corsHeaders ? {
      requestHeaders: this.headers['Access-Control-Allow-Headers'].split(/\s*,\s*/)
    } : null));
  }

  if (options.privateNetworkAccess) {
    this.headers['Access-Control-Allow-Private-Network'] = true;
  }

  if (options.robots) {
    before.push(function (req, res) {
      if (req.url === '/robots.txt') {
        res.setHeader('Content-Type', 'text/plain');
        var robots = options.robots === true
          ? 'User-agent: *\nDisallow: /'
          : options.robots.replace(/\\n/, '\n');

        return res.end(robots);
      }

      res.emit('next');
    });
  }

  if (typeof options.proxyConfig === 'object') {
    var proxy = httpProxy.createProxyServer();
    before.push(function (req, res, next) {
      for (var key of Object.keys(options.proxyConfig)) {
        if (!minimatch(req.url, key)) continue;
        req.proxy ??= {};
        var matchConfig = options.proxyConfig[key];
        
        if (matchConfig.pathRewrite) {
          Object.entries(matchConfig.pathRewrite).forEach(rewrite => {
            req.url = req.url.replace(new RegExp(rewrite[0]), rewrite[1]);
          });
        }
        
        var configEntries = Object.entries(matchConfig).filter(entry => entry[0] !== "pathRewrite");
        configEntries.forEach(entry => req.proxy[entry[0]] = entry[1]);
        break;
      }

      if (req.proxy) {
        if (options.logFn) {
          options.logFn(req, res);
        }
        proxy.web(req, res, req.proxy, function (err, req, res) {
          if (options.logFn) {
            options.logFn(req, res, {
              message: err.message,
              status: res.statusCode });
          }
          res.emit('next');
        });
      } else {
        next();
      }
    });
  }

  if (!proxyAll) {
    before.push(httpServerCore({
      root: this.root,
      baseDir: options.baseDir,
      cache: this.cache,
      showDir: this.showDir,
      showDotfiles: this.showDotfiles,
      hidePermissions: this.hidePermissions,
      autoIndex: this.autoIndex,
      sort: options.sort,
      defaultExt: this.ext,
      dirOverrides404: this.dirOverrides404,
      gzip: this.gzip,
      brotli: this.brotli,
      forceContentEncoding: this.forceContentEncoding,
      contentType: this.contentType,
      mimetypes: options.mimetypes,
      handleError: typeof options.proxy !== 'string'
    }));
  }

  if (typeof options.proxy === 'string') {
    var proxyOptions = options.proxyOptions || {};

    if (proxyOptions.changeOrigin == null) {
        proxyOptions.changeOrigin = true;
    }

    var proxy = httpProxy.createProxyServer({
      ...proxyOptions,
      target: options.proxy,
    });
    before.push(function (req, res) {
      proxy.web(req, res, {}, function (err, req, res) {
        if (options.logFn) {
          options.logFn(req, res, {
            message: err.message,
            status: res.statusCode });
        }
        res.emit('next');
      });
    });
  }

  var serverOptions = {
    before: before,
    headers: this.headers,
    onError: function (err, req, res) {
      if (options.logFn) {
        options.logFn(req, res, err);
      }

      res.end();
    }
  };

  if (options.https) {
    serverOptions.https = options.https;
  }

  this.server = serverOptions.https && serverOptions.https.passphrase
    // if passphrase is set, shim must be used as union does not support
    ? require('./shims/https-server-shim')(serverOptions)
    : union.createServer(serverOptions);

  if (isNaN(options.timeout) || isNaN(parseFloat(options.timeout))) {
    this.server.setTimeout(120);
  } else {
    // set custom timeout only if options.timeout is a numeric string
    this.server.setTimeout(Math.max(0, Number(options.timeout)));
  }

  if (typeof options.proxy === 'string' && options.websocket) {
    this.server.on('upgrade', function (request, socket, head) {
      proxy.ws(request, socket, head, {
        target: options.proxy,
        changeOrigin: true
      }, function (err, req, res) {
        if (options.logFn) {
          options.logFn(req, res, {
            message: err?.message,
            status: res?.statusCode });
        }
        res.emit('next');
      });
    });
  }
}

HttpServer.prototype.listen = function () {
  this.server.listen.apply(this.server, arguments);
};

HttpServer.prototype.close = function () {
  return this.server.close();
};

HttpServer.prototype.address = function () {
  return this.server.address();
};
